packages/flutter/lib/src/material/ink_well.dart

```diff
@@ -30,303 +16,130 @@ class InkWell extends StatefulComponent {
-    this.onLongPress = onLongPress;
-  }
-
-  GestureTapCallback get onTap => _onTap;
-  GestureTapCallback _onTap;
-  void set onTap (GestureTapCallback value) {
-    _onTap = value;
-    _syncTapRecognizer();
-  }
-
-  GestureTapCallback get onDoubleTap => _onDoubleTap;
-  GestureTapCallback _onDoubleTap;
-  void set onDoubleTap (GestureTapCallback value) {
-    _onDoubleTap = value;
-    _syncDoubleTapRecognizer();
-  }
-
-  GestureTapCallback get onLongPress => _onLongPress;
-  GestureTapCallback _onLongPress;
-  void set onLongPress (GestureTapCallback value) {
-    _onLongPress = value;
-    _syncLongPressRecognizer();
-  }
-
-  _HighlightChangedCallback onHighlightChanged;
-
-  final List<_InkSplash> _splashes = new List<_InkSplash>();
-  _InkSplash _lastSplash;
-
-  TapGestureRecognizer _tap;
-  DoubleTapGestureRecognizer _doubleTap;
-  LongPressGestureRecognizer _longPress;
-
-  void _removeSplash(_InkSplash splash) {
-    _splashes.remove(splash);
-    if (_lastSplash == splash)
-      _lastSplash = null;
-  }
 
-  void handleEvent(InputEvent event, BoxHitTestEntry entry) {
-    if (event.type == 'pointerdown' && (onTap != null || onDoubleTap != null || onLongPress != null)) {
-      _tap?.addPointer(event);
-      _doubleTap?.addPointer(event);
-      _longPress?.addPointer(event);
-    }
-  }
-
-  void attach() {
-    super.attach();
-    _syncTapRecognizer();
-    _syncDoubleTapRecognizer();
-    _syncLongPressRecognizer();
-  }
-
-  void detach() {
-    _disposeTapRecognizer();
-    _disposeDoubleTapRecognizer();
-    _disposeLongPressRecognizer();
-    super.detach();
-  }
-
-  void _syncTapRecognizer() {
-    if (onTap == null && onDoubleTap == null && onLongPress == null) {
-      _disposeTapRecognizer();
-    } else {
-      _tap ??= new TapGestureRecognizer(router: FlutterBinding.instance.pointerRouter)
-        ..onTapDown = _handleTapDown
-        ..onTap = _handleTap
-        ..onTapCancel = _handleTapCancel;
-    }
-  }
-
-  void _disposeTapRecognizer() {
-    _tap?.dispose();
-    _tap = null;
-  }
-
-  void _syncDoubleTapRecognizer() {
-    if (onDoubleTap == null) {
-      _disposeDoubleTapRecognizer();
-    } else {
-      _doubleTap ??= new DoubleTapGestureRecognizer(router: FlutterBinding.instance.pointerRouter)
-        ..onDoubleTap = _handleDoubleTap;
-    }
-  }
-
-  void _disposeDoubleTapRecognizer() {
-    _doubleTap?.dispose();
-    _doubleTap = null;
-  }
-
-  void _syncLongPressRecognizer() {
-    if (onLongPress == null) {
-      _disposeLongPressRecognizer();
+  InkHighlight _lastHighlight;
+  Set<InkSplash> _splashes;
+  InkSplash _currentSplash;
+
+  void updateHighlight(bool value) {
+    if (value == (_lastHighlight != null && _lastHighlight.active))
+      return;
+    if (value) {
+      if (_lastHighlight == null) {
+        RenderBox referenceBox = context.findRenderObject();
+        assert(Material.of(context) != null);
+        _lastHighlight = Material.of(context)?.highlightRectAt(
```

Comment:
Why do you both assert that it's non-null and then handle the case of it being null?